### PR TITLE
Improve coredump_dir on FreeBSD and Solaris based OS

### PR DESCRIPTION
**Pull request by devnexen**
Disclose that coredump_dir may also make the process "traceable",
including enabling core dumps and ptrace(2) attachments in certain
environments.

Add support for making the process traceable in FreeBSD- and
Solaris-like environments.

When Squid does not know how to make the process traceable in a given
environment, it does nothing under the assumption that the process is
already traceable enough (by default) to dump cores. Alternatively, we
could warn the admin about the lack of tracing support for that
environment. It is not clear which option is better, but we are betting
on the processes being traceable by default in most not explicitly
covered environments.

### CODE REVIEW

File: src/comm/Tcp.cc [hunk 1]
```diff
@@ -15,6 +15,9 @@
 #if HAVE_NETINET_TCP_H
 #include <netinet/tcp.h>
 #endif
+#if HAVE_NETINET_IN_H
+#include <netinet/in.h>
+#endif
```
Comment (yadij) [inline src/comm/Tcp.cc:20]: This change also does not appear to be related to the rest of this patch. If this is a second issue being fixed please submit it separately.

File: compat/proc.h [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+#endif
+#if HAVE_SYS_PROCCTL_H
+#include <sys/procctl.h>
+#endif
+#if HAVE_PRIV_H
+#include <priv.h>
+#endif
+
+inline int
+ProcSetTraceable() {
```
Comment (yadij) [inline compat/proc.h:23]: This function does not match up with the intended use of libcompat as a substitute source of missing API symbols. eg PR_SET_DUMPABLE and prctl() themselves in this case.

I like your implementation better than what would be needed to fit here in libcompat. So my request is that you move this to a static function in src/tools.cc instead of extending libcompat.

**Review by yadij — CHANGES_REQUESTED** [review]

File: src/CpuAffinitySet.cc [hunk 3]
```diff
@@ -38,7 +38,7 @@ CpuAffinitySet::apply()
     } else {
         cpu_set_t cpuSet;
         memcpy(&cpuSet, &theCpuSet, sizeof(cpuSet));
-        (void) CPU_AND(&cpuSet, &cpuSet, &theOrigCpuSet);
+        CPU_AND(&cpuSet, &cpuSet, &theOrigCpuSet);
```
[the remark this answers is not in the document]
Comment (devnexen) [reply src/CpuAffinitySet.cc:41]: I know but it does not build otherwise and not sure how fast it would be merged if I do a separate PR.

File: src/comm/Tcp.cc (hunk 1 above)
Comment (devnexen) [reply src/comm/Tcp.cc:20]: ditto.

File: compat/proc.h (hunk 2 above)
Comment (devnexen) [reply compat/proc.h:23]: ok.

File: src/CpuAffinitySet.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (yadij) [reply src/CpuAffinitySet.cc:41]: That is not a valid reason to combine changes.

Yes, the QA work to figure out why that ```(void)``` exists and test on OS other than yours will take some amount of time. That has to be done to accept this change regardless of which PR it is part of. Merging several changes just slows the whole bunch to the speed of the most controversial part.

**Comment by yadij** [comment]
Marking this PR as a feature addition since it adds previously unsupported/missing core dumping. It will need a doc/release/release-6.sgml note about the squid.conf coredup_dir behaviour change.

**Review by yadij — APPROVED** [review]
LGTM.

**Comment by yadij** [comment]
OK to test

**Review by rousskov — CHANGES_REQUESTED** [review]
Thank you for improving support on these platforms. Just a few polishing and documentation requests from me. Most are optional. Please let me know if you prefer that I enact them myself.

> Enable coredump_dir on FreeBSD and Solaris

The documented, primary functionality of `coredump_dir` is already "enabled" on those OSes -- it is not OS-specific. Please adjust the title to something like "Improve coredump_dir support on ..."

**Comment by yadij** [comment]
Actually, kicking this back to @devnexen since the [requested](https://github.com/squid-cache/squid/pull/974#discussion_r800069544) src/cf.data.pre text is still missing even though GH says that request is outdated

File: src/cf.data.pre [hunk 4]
```diff
@@ -5639,6 +5639,13 @@ DOC_START
 	that exists, Squid will chdir() to that directory at startup
 	and coredump files will be left there.
 
+	In addition to changing the directory, the process permissions are updated
+	to enable process tracing and/or coredump file generation. The details are
+	OS-specific, but look for prctl(2) PR_SET_DUMPABLE and procctl(2)
+	PROC_TRACE_CTL documentation as guiding examples. If Squid does not know
+	how to update relevant process permissions in your build environment, it
+	logs an ENOSYS-based ERROR message.
```
Comment (rousskov) [inline src/cf.data.pre:5647]: It feels wrong to log an ERROR message when Squid did not fail to update permissions. In the current `return ENOSYS` case, Squid does not even know whether such an update is necessary in the current environment! Should not that be a WARNING (or just a "for your information" status/progress message) instead?

**Review by rousskov — CHANGES_REQUESTED** [review]
PR code just needs minor fixes/polishing, but I would like to resolve one question _before_ changing the code. Please see the inlined change request for details.

I also committed two documentation updates. Please make sure you are OK with them (or adjust the PR further as needed to make you happy).

File: src/cf.data.pre (hunk 4 above)
Comment (devnexen) [reply src/cf.data.pre:5647]: I find a ERROR a bit "overreacting" indeed, this is more a icing on a cake.

Comment (rousskov) [reply src/cf.data.pre:5647]: Glad we agree! I will wait for @yadij opinion before proceeding with the final(?) adjustments.

Comment (yadij) [reply src/cf.data.pre:5647]: In earlier review I assumed there was some importance you knew of to bump the message up to level DBG_IMPORTANT and label as ERROR.

If this is just a nice-to-have behaviour leaving it as level-2 with a WARNING should be sufficiently high. I do not have a preference between WARNING or NOTICE.

Comment (rousskov) [reply src/cf.data.pre:5647]: > In earlier review I assumed there was some importance you knew of to bump the message up to level DBG_IMPORTANT and label as ERROR.

Your assumption was correct. In fact, I explicitly [stated](https://github.com/squid-cache/squid/pull/974#discussion_r800069683) the exact reason -- we should let the admin know (that there was a failure in applying their configuration directive). I do not consider level-2+ messages as "letting admin know". Level-2+ messages can be essential in triage, but they do not work well as admin notifications. Level-2+ is "pull", level-0/1 can be considered "push". Configuration failures deserve push notifications.

> If this is just a nice-to-have behaviour leaving it as level-2 with a WARNING should be sufficiently high.

Leaving failure reporting at level-2 does not address the [earlier](https://github.com/squid-cache/squid/pull/974#discussion_r800069683) concern, but this thread has enough information for me to be able to polish this. I will be back with specific changes shortly.

>  I do not have a preference between WARNING or NOTICE.

Whenever possible, for problem messages, we should be using one of the three standardized labels: `FATAL:`, `ERROR:`, `WARNING:`. See d816f28 for more information. That commit was about level-0/1 messages, but it is easy to adjust its rationale to become applicable to level-2 problem messages.

Comment (rousskov) [reply src/cf.data.pre]: > this thread has enough information for me to be able to polish this. I will be back with specific changes shortly.

Done. The primary changes are in 44c7c4d. I consider this change request addressed.

File: src/tools.cc [hunk 5]
```diff
@@ -273,6 +279,49 @@ rusage_pagefaults(struct rusage *r)
...
+setTraceability()
+{
+    // for now, setting coredump_dir is required to make the process traceable
+    if (!Config.coredump_dir)
+        return;
+
+    try {
+        makeTraceable();
+    } catch (...) {
+        debugs(50, DBG_IMPORTANT, "ERROR: Cannot make the process traceable:" <<
```
Comment (devnexen) [inline src/tools.cc:320]: There can be some mixing messages, having WARNING and ERROR in same time ?

**Review by rousskov — APPROVED** [review]
I have adjusted PR code and description. See commit 44c7c4d for primary details.  Here is a sample cache.log error message:

    2022/02/09 12:14:29| ERROR: Cannot make the process traceable:
        exception: prctl(PR_SET_DUMPABLE) failure: (1) Operation not permitted
        exception location: tools.cc(289) operator()

Please review and adjust further if needed, but I am now happy to see this PR merged.

@devnexen, thank you for improving this code and exposing an important coredump_dir side effect!

File: src/tools.cc (hunk 5 above)
Comment (rousskov) [reply src/tools.cc:320]: No, the level-2 `WARNING` code does not throw and, hence, is not going to trigger this level-1 `ERROR` message.

Comment (devnexen) [reply src/tools.cc:320]: ok

File: src/tools.cc [hunk 6]
```diff
@@ -273,6 +279,50 @@ rusage_pagefaults(struct rusage *r)
...
+    // possibly save one getpid syscall, for now still necessary.
+    int traceable = PROC_TRACE_CTL_ENABLE;
+    if (procctl(P_PID, getpid(), PROC_TRACE_CTL, &traceable) != 0)
+        handleError("procctl(PROC_TRACE_CTL_ENABLE)", errno);
+#elif HAVE_SETPFLAGS
+    if (setpflags(__PROC_PROTECT, 0) != 0)
+        handleError("setpflags(__PROC_PROTECT)", errno);
+#else
+    debugs(50, 2, "WARNING: Assuming this process is traceable");
+    (void)handleError; // just "use" the variable; there is no error here
```
Comment (devnexen) [inline src/tools.cc:305]: [[maybe_unused]] would not work ?

Comment (rousskov) [reply src/tools.cc:305]: It [will](https://en.cppreference.com/w/cpp/language/attributes/maybe_unused), but only after we upgrade to C++17. For now, Squid uses C++11 and `(void)`.

Comment (rousskov) [reply src/tools.cc:305]: And normally we do not add comments for such `(void)x` statements -- they are considered self-documenting. This one is a special case, as I tried to explain in 07b45b0.

Comment (devnexen) [reply src/tools.cc:305]: ah right I forgot it s only C++11...

**Comment by rousskov** [comment]
@yadij, this PR is now on your side: There were significant changes after your earlier approval, so I do not want to merge it without giving you a chance to look over the current PR code/description...

**Review by yadij — APPROVED** [review]
